include/drivers/dai.h

```diff
@@ -0,0 +1,629 @@
+	/** @brief Discard the transmit / receive queue.
+	 *
+	 * Stop the transmission / reception immediately and discard the
+	 * contents of the respective queue. This trigger can be used in any
+	 * state other than NOT_READY and changes the interface state to READY.
+	 */
+	DAI_TRIGGER_DROP,
+	/** @brief Prepare the queues after underrun/overrun error has occurred.
+	 *
+	 * This trigger can be used in ERROR state only and changes the
+	 * interface state to READY.
+	 */
+	DAI_TRIGGER_PREPARE,
+	/** @brief Reset
+	 *
+	 * This trigger frees resources and moves the driver back to initial
+	 * state as in driver init.
+	 */
+	DAI_TRIGGER_RESET,
+};
+
+/** \brief Properties of DAI
+ *
+ * This struct is used with APIs get_properties function to query DAI
+ * properties like fifo address and dma handshake. These are needed to
+ * for example to setup dma outside driver code.
+ */
+struct dai_properties {
+	uint32_t fifo_address; /* fifo address */
+	uint32_t dma_hs_id; /* dma handshake id */
+	uint32_t reg_init_delay; /* delay for register init */
+};
+
+/** Main dai config struct */
+struct dai_config {
+	enum dai_type type;
+	uint32_t dai_index;
+
+	/* generic dai config members */
+	uint8_t channels;
+	dai_fmt_t format;
+	dai_opt_t options;
+	uint32_t frame_clk_freq;
+
+	/* optional dai specific params */
+	const void *params;
+};
+
+/* TODO 1) array of clocks 2) wallclk naming to be clear */
+struct dai_ts_cfg {
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+	int type; /* SSP, DMIC, HDA, etc. */
+	int direction; /* Playback, capture */
+	int index; /* For SSPx to select correct timestamp register */
+	int dma_id; /* DMA instance id */
+	int dma_chan_index; /* Used DMA channel */
+	int dma_chan_count; /* Channels in single DMA */
+};
+
+struct dai_ts_data {
+	uint64_t walclk; /* Wall clock */
+	uint64_t sample; /* Sample count */
+	uint32_t walclk_rate; /* Rate in Hz, e.g. 19200000 */
+};
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * For internal use only, skip these in public documentation.
+ */
+__subsystem struct dai_driver_api {
+	int (*init)(const struct device *dev);
+	int (*deinit)(const struct device *dev);
+
+	int (*config_set)(const struct device *dev, const struct dai_config *cfg,
+			  const void * bespoke_cfg);
+	const struct dai_config *(*config_get)(const struct device *dev,
+					       enum dai_dir dir);
+
+	int (*get_properties)(const struct device *dev, enum dai_dir dir, int stream_id,
+			      struct dai_properties *prop);
+
+	int (*trigger)(const struct device *dev, enum dai_dir dir, int stream_id,
+		       enum dai_trigger_cmd cmd);
+
+	int (*read)(const struct device *dev, void *buf, size_t *size);
+	int (*write)(const struct device *dev, void *buf, size_t size);
+
+	/* timestamp ops - optional */
+	int (*ts_config)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_start)(const struct device *dev, struct dai_ts_cfg *cfg);
+	int (*ts_stop)(const struct device *dev, struct dai_ts_cfg *cfg);
```

Comment:
Could you clarify start and stop flow? What about async operation support w/ callback? (might be useful for hw implementation that may take a while to capture the correlated ts).

Reply:
@mmaka1  This is basicly documented in detail here:
https://github.com/thesofproject/sof/blob/e5d95d1dd3b336f3c465c137187d820e427c00b4/src/include/sof/audio/component.h#L384
Basicly this allows explicit control of timestamp generation in the DAI. Example implementation in a SOF DAI driver for a I2S/SSP: https://github.com/thesofproject/sof/blob/e5d95d1dd3b336f3c465c137187d820e427c00b4/src/drivers/intel/cavs/timestamp.c#L237

@juimonen  Maybe you can add the SOF component.h documentation for these methods here? For some reason SOF documentation for this is in component header, not the DAI header.